invertedai/large/drive.py

```diff
@@ -0,0 +1,314 @@
+        Optionally specify the version of the model. If None is passed which is by default, 
+        the best model will be used.
+    
+    capacity:
+        The number of agents allowed in a region before it must subdivide. Currently this 
+        value represents the capacity of a quadtree leaf node that will subdivide if the 
+        number of vehicles in the region passes this threshold.
+
+    is_async:
+        A flag to control whether to use asynchronous DRIVE calls.
+
+    See Also
+    --------
+    :func:`drive`
+    """
+    
+    num_agents = len(agent_states)
+    agent_x = [None]*num_agents
+    agent_y = [None]*num_agents
+    for i, agent in enumerate(agent_states):
+        agent_x[i] = agent.center.x
+        agent_y[i] = agent.center.y
+    max_x, min_x, max_y, min_y = max(agent_x), min(agent_x), max(agent_y), min(agent_y)
+    region_size = ceil(max(max_x,max_y) - min(min_x,min_y))
+    region_center = ((max_x+min_x)/2,(max_y+min_y)/2)
+
+    quadtree = QuadTree(
+        capacity=capacity,
+        region=Region.init_square_region(
+            Point.fromlist(list(region_center)),
+            region_size
+        ),
+    )
+    for i, (agent, attrs, recurr_state) in enumerate(zip(agent_states,agent_attributes,recurrent_states)):
+        agent_info = AgentInfo.fromlist([agent, attrs, recurr_state, i])
+        is_inserted = quadtree.insert(agent_info)
+
+    all_leaf_nodes = quadtree.get_leaf_nodes()
+    async_input_params = []
+    all_responses = []
+    non_empty_regions = []
+    agent_id_order = []
+    for i, leaf_node in enumerate(all_leaf_nodes):
+        region, region_buffer = leaf_node.region, leaf_node.region_buffer
+        region_agents_ids = [particle.agent_id for particle in leaf_node.particles]
+        
+        if len(region.agent_states) > 0:
+            non_empty_regions.append(region)
+            agent_id_order.extend(region_agents_ids)
+            input_params = {
+                "location":location,
+                "agent_attributes":region.agent_attributes+region_buffer.agent_attributes,
+                "agent_states":region.agent_states+region_buffer.agent_states,
+                "recurrent_states":region.recurrent_states+region_buffer.recurrent_states,
+                "light_recurrent_states":light_recurrent_states,
+                "traffic_lights_states":traffic_lights_states,
+                "get_birdview":get_birdview,
+                "rendering_center":rendering_center,
+                "rendering_fov":rendering_fov,
+                "get_infractions":get_infractions,
+                "random_seed":random_seed,
+                "api_model_version":api_model_version
+            }
+            if not is_async:
+                all_responses.append(iai.drive(**input_params))
+            else:
+                async_input_params.append(input_params)
+
+    if is_async:
+        all_responses = asyncio.run(async_drive_all(async_input_params))
+
+    response = DriveResponse(
```

Comment:
This whole `_flatten_and_sort` business is quite convoluted and could use refactoring, but we can leave that for later.

Reply:
It is an annoying circumstance for sure. There is no connection between the location of the agent and it's index in the input parameter. This is a situation where an explicit ID for each agent would be useful but I understand it wouldn't make sense to change one of our basic data classes just to solve this. 

I think a middle ground is that this flatten and sort method can be a member function of the quadtree class instead to make it a bit cleaner remaining functionally the same.